src/java.desktop/unix/native/libawt/awt/awt_LoadLibrary.c

```diff
@@ -61,16 +67,13 @@ JNIEXPORT jboolean JNICALL AWTIsHeadless() {
         env = (JNIEnv *)JNU_GetEnv(jvm, JNI_VERSION_1_2);
         graphicsEnvClass = (*env)->FindClass(env,
                                              "java/awt/GraphicsEnvironment");
-        if (graphicsEnvClass == NULL) {
-            return JNI_TRUE;
-        }
+        CHECK_EXCEPTION_FATAL(env, "java/awt/GraphicsEnvironment class not found");
```

Comment:
Taking out any discussion about graal/etc as well as absent of any classes and methods, I agree that the current code is questionable.

But the solution to always print "java/awt/GraphicsEnvironment class not found" does not seems to be correct as well.
 
One of the problem with this code is an actual conflict between two patches. Both are related to warning caused by pending java exceptions, but implemented differently:
 - https://bugs.openjdk.org/browse/JDK-8031001: treated a java exceptions as a fatal errors and fail fast
 - https://bugs.openjdk.org/browse/JDK-8130507: ignores  a java exceptions and tried to fallback into the headless mode in assumption that it may be fine for the application

Both of that patches only changes the code paths which were reported by "some tools"/checkjni - this is the reason why the next code was not reported and was not patched(which is wrong):

```
       graphicsEnvClass = (*env)->FindClass(env,
                                             "java/awt/GraphicsEnvironment");
        if (graphicsEnvClass == NULL) {
            return JNI_TRUE;
        }
        headlessFn = (*env)->GetStaticMethodID(env,
                                               graphicsEnvClass, "isHeadless", "()Z");
        if (headlessFn == NULL) {
            return JNI_TRUE;
        }
```
I we would like to follow JDK-8031001 approach we should use fatal error here and fail fast, for example if OOM is occurred.
I we would like to follow JDK-8130507 we should clear an exceptions and try to use headless mode.

I guess ignoring an exceptions is not that good and the fix for JDK-8130507 should be rethinking.

Reply:
You can change the CHECK_EXCEPTION_FATAL macro to something like
```
#define CHECK_EXCEPTION_FATAL(env, message) \
    if ((*env)->ExceptionCheck(env)) { \
        (*env)->ExceptionDescribe(env);
        (*env)->FatalError(env, message); \
    }
```
In this case the root cause of the bug will always be printed.
And then update the fatal message to some generic test.

Note that the FatalError is used in this code since we always should load the library(libawt_xawt or libawt_headless) or fail fast, otherwise we most probably will get an error later.

Reply:
> You can change the CHECK_EXCEPTION_FATAL macro to something like
> 
> ```
> #define CHECK_EXCEPTION_FATAL(env, message) \
>     if ((*env)->ExceptionCheck(env)) { \
>         (*env)->ExceptionDescribe(env);
>         (*env)->FatalError(env, message); \
>     }
> ```
> 
> In this case the root cause of the bug will always be printed. And then update the fatal message to some generic text.
> 
> Note that the FatalError is used in this code since we always should load the library(libawt_xawt or libawt_headless) or fail fast, otherwise we most probably will get an error later.

Hello @mrserb,  would this be O.K. then?  [91d554f](https://github.com/openjdk/jdk/pull/20169/commits/91d554f4919a9923a304ec1d3b23b96cb5cf144a)